src/App.js

```diff
@@ -23,49 +25,107 @@ const generateSquares = () => {
   }
 
   return squares;
-}
+};
 
 const App = () => {
-
   // This starts state off as a 2D array of JS objects with
   // empty value and unique ids.
+
   const [squares, setSquares] = useState(generateSquares());
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
+  const [winner, setWinner] = useState(null);
+
+  const onClickCallback = (id) => {
+    setSquares((squares) =>{
+      let newBoard = squares.map((square)=>{
+        for (let property of square){
+          if (property.id === id && property.value === ''){
+            if (currentPlayer === PLAYER_1){
+              property.value = PLAYER_1;
+            } else if (currentPlayer === PLAYER_2){
+              property.value = PLAYER_2;
+            }
+          }
+        }
+        return square;
+      });
+    setWinner(checkForWinner());
+    
+    return newBoard;
+    });
+
+    if (currentPlayer === PLAYER_1){
+      setCurrentPlayer(PLAYER_2);
+    }else {
+      setCurrentPlayer(PLAYER_1);
+    }
```

Comment:
What should happen if X plays in the top left, then O tries to play in the top-left as well. Your click handler would prevent O from placing a piece over X, but since we don't track whether the attempted click was a valid play, it still winds up switching to X's turn. So O loses their turn. This isn't the behavior that we want.